src/storage/src/encoding/row_based_deserializer.rs

```diff
@@ -43,9 +45,53 @@ impl Decoding for RowBasedDeserializer {
         value: impl AsRef<[u8]>,
     ) -> Result<Option<(risingwave_common::types::VirtualNode, Vec<u8>, Row)>> {
         // todo: raw_key will be used in row-based pk dudup later.
+        self.deserialize_inner(raw_key, value)
+    }
+}
+
+impl RowSerde for RowBasedDeserializer {
+    type Deserializer = RowBasedDeserializer;
+    type Serializer = RowBasedSerializer;
+
+    fn create_serializer(
+        pk_indices: &[usize],
+        column_descs: &[ColumnDesc],
+        column_ids: &[ColumnId],
+    ) -> Self::Serializer {
+        super::Encoding::create_row_serializer(pk_indices, column_descs, column_ids)
+    }
+
+    fn create_deserializer(
+        column_mapping: std::sync::Arc<super::ColumnDescMapping>,
+        data_types: Vec<risingwave_common::types::DataType>,
+    ) -> Self::Deserializer {
+        Decoding::create_row_deserializer(column_mapping, data_types)
+    }
+}
+impl RowBasedDeserializer {
+    fn deserialize_inner(
+        &mut self,
+        raw_key: impl AsRef<[u8]>,
+        value: impl AsRef<[u8]>,
+    ) -> Result<Option<(VirtualNode, Vec<u8>, Row)>> {
+        let raw_key = raw_key.as_ref();
+        if raw_key.len() < VIRTUAL_NODE_SIZE {
+            // vnode + cell_id
+            return Err(ErrorCode::InternalError(format!(
+                "corrupted key: {:?}",
+                Bytes::copy_from_slice(raw_key)
+            ))
+            .into());
+        }
+
+        let (vnode, key_bytes) = {
+            let (vnode_bytes, key_bytes) = raw_key.split_at(VIRTUAL_NODE_SIZE);
+            let vnode = VirtualNode::from_be_bytes(vnode_bytes.try_into().unwrap());
+            (vnode, key_bytes)
+        };
```

Comment:
Maybe this can be refactored out into separate function for readability and reuse:
```rs
fn parse_raw_key_to_vnode_and_key(raw_key) -> (vnode, bytes)
```

Can be done separately.